### PR TITLE
Secrets: Move license check to secret store plugin

### PR DESCRIPTION
**Pull request by guicaulada**
**What this PR does / why we need it**:

This PR moves the license check part to the plugin itself so in the future we can implement open-source secret store plugins as well.

**Which issue(s) this PR fixes**:

Fixes grafana/grafana-aws-secrets-manager#14

**Special notes for your reviewer**:

Dependent on:
https://github.com/grafana/grafana-enterprise/pull/3672
https://github.com/grafana/grafana-aws-secrets-manager/pull/20

### CODE REVIEW

File: pkg/services/secrets/kvstore/plugin_mig.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/plugin_mig.go:44]: Still need to manually check if the plugin is installed before we can do any migrations. We can pass in the Secrets Plugin Manager to the provider and just check that way.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/kvstore.go:21]: While I am not personally bothered by using a global logger, I have seen PR feedback that seems to be against doing this. Should we just pass the logger into the helper functions instead?

File: pkg/services/secrets/kvstore/kvstore.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/kvstore.go:40]: Based on the conversation in slack earlier, I think we can even do away with this logic and just have a check for the plugin being installed. This would support migrating secrets to and from the plugin without it necessarily being enabled. 

So something like:
```
if plugin installed
    start plugin

if use_plugin = true
   return wrapped remote plugin store
else 
   return wrapped sql store
        
```

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/remote_plugin.go:205]: I do think it is still worth guarding against multiple calls to this function. I would suggest using sync.Once and sticking the start logic in there just in case.

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/remote_plugin.go:191]: See comment on kvstore.go. We could potentially just do away with this all together by just always starting the plugin if installed, and then just using the use_plugin flag for everything else.

**Review by mmandrus — CHANGES_REQUESTED** [review]
Are there some missing files from this PR? Since you deleted the remote plugin check files, I would expect compile errors in the wire_oss.go file and frontendsettings.go (which should just check the use_plugin flag instead of the old remote check).

Otherwise, I think these are positive changes. I think we can further simplify things by just always starting the plugin as discussed in slack. See below.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/kvstore.go:40]: But what did we say about the scenario where the plugin is not installed but the `use_plugin = true` ? in that case the plugin would be nil if Im not wrong

Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go:40]: I believe `shouldUseRemoteSecretsPlugin` already accounts for `if plugin installed && use_plugin else sql_store` in this case this already works like we described on our conversations, I think migrating to the plugin without use_plugin enabled might not be a good approach since the user might not desire to have a bunch of secrets created on his secret store.

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/remote_plugin.go:205]: Will do!

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/remote_plugin.go:191]: Wouldn't it save resources to not start the plugin if it's not going to be used?

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/remote_plugin.go:206]: Done in 10b1f4fea7

File: pkg/services/secrets/kvstore/kvstore.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go:21]: Since there are a lot of functions that are not part of an interface that use `logger` I was thinking it would be best to keep it as a global instead of passing it as an argument to keep the code cleaner.

File: pkg/services/secrets/kvstore/remote_plugin.go
Comment (mmandrus) [reply pkg/services/secrets/kvstore/remote_plugin.go:192]: The case I could see would be if we need to trigger a migration from the plugin later on, like if we add an API call to do it. In the case, we could just make sure the plugin is started at that time.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go]: Done on 1ccdedf433

File: pkg/services/secrets/kvstore/kvstore.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/kvstore.go:54]: It sounds like we do want to crash the app in this case, which honestly makes sense if we are somehow unable to read from the kvstore. In that case, lets remove this // todo

File: pkg/services/secrets/kvstore/plugin_mig.go
Comment (mmandrus) [inline pkg/services/secrets/kvstore/plugin_mig.go:48]: we still need to check the `migrate_to_plugin` flag, right?

**Review by mmandrus — COMMENTED** [review]
Just a couple of cleanup things. There might be a few more things we can change around, but I think it can wait until I have a PR open for migration from the plugin, since we will have the whole picture at that point.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go]: Done on 41b15d3e21

Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go]: Looking more into the fatalErr I noticed it's actually an error to retrieve the key so I thought it was more appropriate to rename it accordingly.

Done on a74006bee7

File: pkg/services/secrets/kvstore/kvstore.go
Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go]: I think using err we can return both a reason and a checkable condition variable, simplifying things, what do you think? Maybe renaming the function would make it seem less odd?

File: pkg/services/secrets/kvstore/plugin_mig.go
Comment (mmandrus) [reply pkg/services/secrets/kvstore/plugin_mig.go]: I don't actually remember that conversation, but it makes sense. Maybe before we wanted this to be a flag in case migration is slow, but since we have plans to speed it up, this could be fine.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (mmandrus) [reply pkg/services/secrets/kvstore/kvstore.go]: That's true, though it feels like a misuse of error to me, since we debug log the error output. Maybe it is just my Java background speaking!

If we rename it though, it could just be `evaluateUseSecretsPlugin()` and I wouldn't feel like it should return a bool. Your call though, I will approve either way 👍

File: pkg/services/secrets/kvstore/plugin_mig.go
Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/plugin_mig.go]: So this means that when use_plugin=true, we will always trigger a migration?
Remember that running the migration means deleting the secrets from the database, so I think that should be explicit to the admin. We would be performing a non-trivial operation.

File: pkg/services/secrets/kvstore/kvstore.go
Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/kvstore.go]: I also feel it is weird to use an error like that (also Java background :) ). 
 
As far as I can see, the returned error on `shouldUseRemoteSecretsPlugin` is only logged, but not returned outside the function or read to see the kind of error. 
What if we change the return value of `shouldUseRemoteSecretsPlugin` to a boolean, and log the error or information needed within that function?
I could also help clean the "if {} else {} " block below which has the "happy path" within the else

Comment (guicaulada) [reply pkg/services/secrets/kvstore/kvstore.go]: The plan was to remove the logger within the function so we don't need a global logger or an argument logger just to serve that purpose of logging the reason for debug purposes.

Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/kvstore.go]: Oh right, sorry I circle back to that issue

**Comment by guicaulada** [comment]
All my enterprise drone builds are failing on plugins-build step due to GRAFANA_API_KEY that must be set, not sure if that is related to my PR or a change on the pipeline.

**Review by mmandrus — DISMISSED** [review]

File: pkg/services/secrets/kvstore/plugin_mig.go
Comment (mmandrus) [reply pkg/services/secrets/kvstore/plugin_mig.go]: Approved initially but remembered this point and dismissed my review. 

With the PR I'm working on, we'll be migrating to or from the plugin. 

One idea: instead of having `migrate_to_plugin`, we could head this off by having a config setting called `migration_strategy` with the options:
`to_plugin` (default)
`from_plugin`
`none`

This way, by default we have the implied usage, but we give admins the options of keeping old secrets around. Thoughts? @leandro-deveikis @guicaulada

Comment (guicaulada) [reply pkg/services/secrets/kvstore/plugin_mig.go]: @mmandrus just to be sure, in this case we would have a new variable under [secrets] ?

For example:
```
[secrets]
use_plugin = true
migration_strategy = to_plugin
```

Comment (leandro-deveikis) [reply pkg/services/secrets/kvstore/plugin_mig.go]: > One idea: instead of having migrate_to_plugin, we could head this off by having a config setting called [migration_strategy] with the following options mapping to enums:
to_plugin (default)
from_plugin
none

I think that could work, as I said in the chat, we will need some good documentation here

Comment (mmandrus) [reply pkg/services/secrets/kvstore/plugin_mig.go]: @guicaulada that is what I had in mind

**Review by mmandrus — APPROVED** [review]
LGTM!

**Comment by guicaulada** [comment]
This fixes grafana/grafana-partnerships-team#266

**Pull request merged**
